components/builder-depot/src/server.rs

```diff
@@ -1257,24 +1260,17 @@ fn download_package(req: &mut Request) -> IronResult<Response> {
 
     match route_message::<OriginPackageGet, OriginPackage>(req, &ident_req) {
         Ok(package) => {
-            let temp_path = depot.archive_path(package.get_ident(), &target);
+            let dir = tempdir().expect("Unable to create a tempdir!");
```

Comment:
Hmm yeah that's a good point - you may have to use a different pattern - eg, see https://github.com/habitat-sh/habitat/blob/master/components/common/src/ui.rs#L541

Reply:
Your assumption about the variable being in scope is correct - it's in scope for the duration of the match block.  The problem I was referring to involves this sequence of events:
* Client requests hart file
* Depot downloads hart file from S3 and stores it in temp dir
* Depot constructs an HTTP response, including headers for content disposition and file name of the hart file (note that the body of the HTTP response does not contain the contents of the hart file at this point)
* Depot returns the HTTP response
* Immediately after the HTTP response is returned from this function, the temp dir goes out of scope and is deleted

At this point the client has an HTTP response that points to a file that doesn't exist any longer.  I'm not certain of the mechanism by which the client and/or server takes the HTTP response, recognizes that it's supposed to contain a file and puts that file contents into the body of the HTTP response, but it's not present in our code, that I can see.

I think the safer, albeit more laborious, approach is similar to what Salim linked, where we generate a directory with a UUID for a name, drop the file from S3 in there, and download from there.  At some point, we need a way to clean that up, but honestly I don't think that has to be part of this PR.  I'm imagining some kind of background thread that watches for those directories and removes them after X amount of time.  Right now though, it's sufficient to just leave a comment in the code that stuff isn't getting cleaned up and open an issue to circle back to it.

I think getting the download stuff more bulletproof from a concurrency perspective trumps cleanup at this point.

Reply:
Ok, sorry for all the questions I just want to make sure I have a full understanding - I _thought_ that putting the body of the file into the http response was happening here:
https://github.com/habitat-sh/builder/blob/c463dc8df24af89fec1e9c3510c1d2e15a4e18c9/components/builder-depot/src/server.rs#L1270

We validate that the file exists in the tempdir, and then we pass the archive through to the response function, I just kind of figured since this was inside the match block that this would all be occurring before dropping out of scope:
https://github.com/habitat-sh/builder/blob/c463dc8df24af89fec1e9c3510c1d2e15a4e18c9/components/builder-depot/src/server.rs#L2532

I'd prefer to not go the path of generating a dir based on uuid from the name of the file because of the situation @fnichol was detailing during the outage of like having multiple requests for the same thing and potentially clobbering each other. But we could definitely just use a more durable tempdir type that doesn't get removed unless it's explicitly closed in the code. But then we could end up filling that disk up realllllllly fast.

Reply:
Oh I think I misunderstood your comment, sorry about that @raskchanky I see what you're saying.